### PR TITLE
Put nginx in front of the services

### LINKED ISSUE #50: Put nginx in front of s-pipes-editor-ui and other services
It would be easier to debug s-pipes-editor-ui if we have nginx in front of all services as done in https://github.com/kbss-cvut/record-manager-ui/blob/main/deploy/internal-auth/docker-compose.yml

A/C:
- [x] arguments are given here why nginx in front of docker services is better (maybe @blcham should do it)
- [x] it is implemented
- [x] it is properly described how to use it in https://github.com/kbss-cvut/s-pipes-editor-ui?tab=readme-ov-file#development-with-docker-compose

### PR DESCRIPTION
**Pull request by palagdan**
Resolves #50

### CODE REVIEW

**Comment by palagdan** [comment]
@blcham 

Could you try running this configuration? I tried it on my computer, but I keep getting the same error, even after switching to the main branch and running spe.sh. I also tried removing the cache and containers, but nothing happens. It worked before with the same docker-compose on the main branch.

Screenshot from 2024-09-09 16-08-37

**Comment by palagdan** [comment]
@blcham 

All services seems to be working with nginx except for db-server. I tried to fix it, but I don't know what the problem. 
Screenshot from 2024-09-09 17-14-30

**Comment by palagdan** [comment]
@blcham 

Screenshot from 2024-09-17 20-38-01
he backend is working correctly. I might have requested the wrong path earlier.

File: README.md [hunk 1]
```diff
@@ -115,16 +116,16 @@ where <SPIPES_BACKEND_URL> denotes the URL where SPipes backend is running.
 ### Docker-compose
 
 You can run editor together with backend using docker orchestration. The docker-compose is composed of 4 services and can be run via `docker-compose up`:
-* [chlupnoha/s-pipes-editor-ui:latest](https://hub.docker.com/repository/docker/chlupnoha/s-pipes-editor-ui) - accessible on `http://localhost:3000`
-* [chlupnoha/s-pipes-editor-rest:latest](https://hub.docker.com/repository/docker/chlupnoha/s-pipes-editor-rest) - accessible on `http://localhost:18115` with [configuration options](https://github.com/kbss-cvut/s-pipes-editor#dockerization)
-* [chlupnoha/spipes-engine:latest](https://hub.docker.com/repository/docker/chlupnoha/spipes-engine) - accessible on `http://localhost:8081` with [configuration options](https://github.com/kbss-cvut/s-pipes#dockerization) 
-* [eclipse/rdf4j-workbench:amd64-3.5.0](https://hub.docker.com/r/eclipse/rdf4j-workbench) - accessible on `http://localhost:8080/rdf4j-workbench`
+* [chlupnoha/s-pipes-editor-ui:latest](https://hub.docker.com/repository/docker/chlupnoha/s-pipes-editor-ui) - accessible on `http://localhost:1235`
```
Comment (blcham) [inline README.md:119]:
```suggestion
* [kbss-cvut/s-pipes-editor-ui:latest](https://ghcr.io/kbss-cvut/s-pipes-editor/s-pipes-editor-ui:latest) - accessible on `http://localhost:1235`
```

File: README.md [hunk 2]
```diff
@@ -115,16 +116,16 @@ where <SPIPES_BACKEND_URL> denotes the URL where SPipes backend is running.
...
+* [chlupnoha/s-pipes-editor-rest:latest](https://hub.docker.com/repository/docker/chlupnoha/s-pipes-editor-rest) - accessible on `http://localhost:1235/rest` with
```
Comment (blcham) [inline README.md:120]:
```suggestion
* [kbss-cvut/s-pipes-editor-rest:latest](https://ghcr.io/kbss-cvut/s-pipes-editor/s-pipes-editor:latest) - accessible on `http://localhost:1235/rest` with 
```

File: README.md [hunk 3]
```diff
@@ -115,16 +116,16 @@ where <SPIPES_BACKEND_URL> denotes the URL where SPipes backend is running.
...
-* [chlupnoha/s-pipes-editor-rest:latest](https://hub.docker.com/repository/docker/chlupnoha/s-pipes-editor-rest) - accessible on `http://localhost:18115` with [configuration options](https://github.com/kbss-cvut/s-pipes-editor#dockerization)
-* [chlupnoha/spipes-engine:latest](https://hub.docker.com/repository/docker/chlupnoha/spipes-engine) - accessible on `http://localhost:8081` with [configuration options](https://github.com/kbss-cvut/s-pipes#dockerization) 
-* [eclipse/rdf4j-workbench:amd64-3.5.0](https://hub.docker.com/r/eclipse/rdf4j-workbench) - accessible on `http://localhost:8080/rdf4j-workbench`
+* [chlupnoha/s-pipes-editor-ui:latest](https://hub.docker.com/repository/docker/chlupnoha/s-pipes-editor-ui) - accessible on `http://localhost:1235`
+* [chlupnoha/s-pipes-editor-rest:latest](https://hub.docker.com/repository/docker/chlupnoha/s-pipes-editor-rest) - accessible on `http://localhost:1235/rest` with 
+* [chlupnoha/spipes-engine:latest](https://hub.docker.com/repository/docker/chlupnoha/spipes-engine) - accessible on `http://localhost:1235/services/s-pipes`
```
Comment (blcham) [inline README.md:121]:
```suggestion
* [kbss-cvut/spipes-engine:latest](https://ghcr.io/kbss-cvut/s-pipes-engine/s-pipes-engine:latest) - accessible on `http://localhost:1235/services/s-pipes` 
```

File: README.md [hunk 4]
```diff
@@ -115,16 +116,16 @@ where <SPIPES_BACKEND_URL> denotes the URL where SPipes backend is running.
...
-* [chlupnoha/spipes-engine:latest](https://hub.docker.com/repository/docker/chlupnoha/spipes-engine) - accessible on `http://localhost:8081` with [configuration options](https://github.com/kbss-cvut/s-pipes#dockerization) 
-* [eclipse/rdf4j-workbench:amd64-3.5.0](https://hub.docker.com/r/eclipse/rdf4j-workbench) - accessible on `http://localhost:8080/rdf4j-workbench`
+* [chlupnoha/s-pipes-editor-ui:latest](https://hub.docker.com/repository/docker/chlupnoha/s-pipes-editor-ui) - accessible on `http://localhost:1235`
+* [chlupnoha/s-pipes-editor-rest:latest](https://hub.docker.com/repository/docker/chlupnoha/s-pipes-editor-rest) - accessible on `http://localhost:1235/rest` with 
+* [chlupnoha/spipes-engine:latest](https://hub.docker.com/repository/docker/chlupnoha/spipes-engine) - accessible on `http://localhost:1235/services/s-pipes` 
+* [graphdb](https://github.com/kbss-cvut/s-pipes-editor-ui/tree/master/deploy/db-server) - accessible on `http://localhost:1235/services/db-server`
```
Comment (blcham) [inline README.md:122]:
```suggestion
* [graphdb](./deploy/db-server) - accessible on `http://localhost:1235/services/db-server`
```

File: deploy/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      RDF4J_REPOSITORYNAME: "${RDF4J_REPOSITORYNAME:-s-pipes-hello-world}"
+      RDF4J_PCONFIGURL: "${RDF4J_PCONFIGURL:-./s-pipes-engine/config.ttl}"
+    volumes:
+      - /tmp:/tmp
+      - ${SHARED_ROOT:-/home}:${SHARED_ROOT:-/home}
+      - /usr/local/tomcat/temp/:/usr/local/tomcat/temp/
+
+  s-pipes-engine:
+    image: "ghcr.io/kbss-cvut/s-pipes/s-pipes-engine:latest"
+    #    container_name: s-pipes-engine
```
Comment (blcham) [inline deploy/docker-compose.yml:53]:
```suggestion
```

File: deploy/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    ports:
+      - "8080:8080"
```
Comment (blcham) [inline deploy/docker-compose.yml:55]: We should not need this, but can have:
   expose:
      - "8080"

File: deploy/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,79 @@
+version: "3.9"
+
+services:
+  nginx:
+    image: nginx:latest
+    ports:
+      - "127.0.0.1:${INTERNAL_HOST_PORT:-1235}:80"
+    restart: always
```
Comment (blcham) [inline deploy/docker-compose.yml:8]:
```suggestion
    restart: unless-stopped
```

Comment (blcham) [reply deploy/docker-compose.yml:8]: Use it for all `restart:` sections. We do not want to start editor on restart of computer if we stopped it.

Comment (blcham) [reply deploy/docker-compose.yml:8]: BTW, i do not see `restart:` section anywhere  ... it should be part of each service, please add it there

**Review by blcham — CHANGES_REQUESTED** [review]
See my comments.

File: deploy/nginx/error.html [hunk 8]
```diff
@@ -0,0 +1,35 @@
+<!-- Record manager Nginx reverse proxy-->
```
Comment (blcham) [inline deploy/nginx/error.html:1]:
```suggestion
<!-- Record manager Nginx reverse proxy-->
```

File: README.md (hunk 1 above)
Comment (palagdan) [reply README.md:119]: done

File: README.md (hunk 2 above)
Comment (palagdan) [reply README.md:120]: done

File: README.md (hunk 1 above)
Comment (palagdan) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (palagdan) [reply README.md]: done

Comment (palagdan) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (palagdan) [reply README.md]: done

File: README.md (hunk 4 above)
Comment (palagdan) [reply README.md]: done

File: deploy/docker-compose.yml (hunk 5 above)
Comment (palagdan) [reply deploy/docker-compose.yml]: done

File: deploy/docker-compose.yml (hunk 6 above)
Comment (palagdan) [reply deploy/docker-compose.yml]: done

File: deploy/docker-compose.yml (hunk 7 above)
Comment (palagdan) [reply deploy/docker-compose.yml]: done

File: deploy/nginx/error.html (hunk 8 above)
Comment (palagdan) [reply deploy/nginx/error.html]: done

File: docker-compose.yml
[the remark this answers is not in the document]
Comment (palagdan) [reply docker-compose.yml:1]: Done. I've chagned the Git history to make it more readable and correct.

File: deploy/docker-compose.yml [hunk 9]
```diff
@@ -1,63 +1,79 @@
...
+    restart: unless-stopped
+    depends_on:
+      - s-pipes-engine
+      - s-pipes-editor-rest
+      - db-server
+      - s-pipes-editor-ui
+    environment:
+      NGINX_ENVSUBST_OUTPUT_DIR: "/etc/nginx"
+      APP_ORIGIN: "${PUBLIC_ORIGIN:-http://localhost:${INTERNAL_HOST_PORT:-1235}}"
+      SERVICE_URL: "${SPIPES_EDITOR_REST:-http://s-pipes-editor-rest:18115}"
```
Comment (blcham) [inline deploy/docker-compose.yml:17]: I do not understand role of `SPIPES_EDITOR_REST`, when do you want to use it. Why it is not sufficient to use `http://s-pipes-editor-rest:18115` in nginx directly.

**Comment by blcham** [comment]
@palagdan watch out, i force-pushed (most likely it was not good idea) just to fix formating of nginx.conf

File: deploy/docker-compose.yml (hunk 9 above)
Comment (blcham) [reply deploy/docker-compose.yml:17]: Moreover in nginx.conf you have many services so it does not makes sense to me call it SERVICE_URL but be more specific.

**Review by blcham — CHANGES_REQUESTED** [review]
LGTM, just one comment related SERVICE_URL -- is it somehow used for development or why it is there?

File: deploy/docker-compose.yml (hunk 9 above)
Comment (palagdan) [reply deploy/docker-compose.yml:17]: @blcham
The motivation behind configuring the SPIPER_EDITOR_REST url  was to make it easier to change the API for the frontend, if needed in the future. I can remove it and simply hardcode http://s-pipes-editor-rest:18115 if preferred.

You are also right about SERVICE_URL in nginx.conf—it should be updated to SPIPER_EDITOR_REST_URL to be more speicific.

Comment (blcham) [reply deploy/docker-compose.yml:17]: OK, I saw in README:
```
SPIPES_EDITOR_REST -- to set up different s-pipes-editor backend (defaults to internal docker service s-pipes-editor-rest:18115)
```

Thus, I suggest renaming it **only here** `SPIPES_EDITOR_REST_SERVICE_URL` as in the context of nginx.

Comment (blcham) [reply deploy/docker-compose.yml:17]: @palagdan

**Pull request merged**